webbpsf/optics.py

```diff
@@ -1111,8 +1117,42 @@ def __init__(self, instrument, include_oversize=False, **kwargs):
 
             # Cubic interpolation of of non-uniform 2D grid
             cf = griddata((v2, v3), zvals, (v2_tel, v3_tel), method='cubic').tolist()
+            # Want to perform extrapolation if outside of bounds
             if np.isnan(cf):
-                cf = self.row[zkey]
+                #cf = self.row[zkey]
+
+                # Create fine mesh grid
+                dstep = 1. / 60. # 1" steps
+                xgrid = np.arange(v2.min(), v2.max()+dstep, dstep)
+                ygrid = np.arange(v3.min(), v3.max()+dstep, dstep)
+                X, Y = np.meshgrid(xgrid,ygrid)
+
+                # Cubic interpolation of all points
+                zgrid = griddata((v2, v3), zvals, (X, Y), method='cubic')
+    
+                # There will be some NaNs along the border that need to be replaced
+                ind_nan = np.isnan(zgrid)
+                # Remove rows/cols 1 by 1 until no NaNs
```

Comment:
This part seems kind of ad hoc. In particular it doesn't account for any significantly non-square border (i.e. you'd ideally exclude different numbers of lines in X and Y). That may not matter in practice sufficiently to do anything different - I'm just asking.

Reply:
I agree it's certainly not ideal, especially for MIRI and NIRSpec. I'm open to suggestions of a more optimal way to trim the rows/columns with NaNs (or a better way to perform the extrapolation). I've investigated a number of methods, and this was the best I could come up with for such sparsely samples data. For NIRCam and NIRISS WFE maps, it's fairly robust and quick. Below is an image showing the resulting extrapolation to the V2/V3 limits. I'm a little wary of NIRSpec and MIRI results beyond. Perhaps choosing the closest field point for those two instruments would be better?

![si_wfe_extrap_Z4](https://user-images.githubusercontent.com/17557291/54401761-404f1f80-4686-11e9-9e79-049041588bf2.png)

Reply:
Excellent plot - that really helps visualize things. Agrees that NIRSpec is a bit of a mess with that method. 

One relatively easy thing to try for NIRSpec and MIRI would be to apply a rotation to the X and Y grids, so the grid axes are more closely aligned with the SI FOVs. That would make the mechanics of the trimming work better for those, I think. Then do all the extrapolation and replacement in that rotated grids.